src/component.ts

```diff
@@ -413,19 +410,93 @@ export function viewObserve<A>(
+export type Child<O = any, A = any> =
+  | [CE]
+  | [CE, CE]
+  | [CE, CE, CE]
+  | [CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE]
+  | [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE] // 12
+  | CE<O, A>;
+
 // A dummy interface is required since TypeScript doesn't handle
 // recursive type aliases
 // See: https://github.com/Microsoft/TypeScript/issues/3496#issuecomment-128553540
-export interface ChildList extends Array<Child> {}
+export interface ChildList extends Array<CE> {}
+
+/**
+ * Takes a component type and returns the explicit output of the component.
+ */
+export type ComponentExplicitOutput<C> = C extends Component<infer O, any> ? O : never;
+
+/**
+ * Takes a component type and returns the output of the component.
+ */
+export type ComponentOutput<C> = C extends Component<any, infer A> ? A : never;
+
+type Co<O> = Component<O, any>;
+
+export type ChildExplicitOutput<Ch extends Child> = ComponentExplicitOutput<
+  ToComponent<Ch>
+>;
+
+// Merge component
+export type MC<C1 extends CE, C2 extends CE> = Component<
+  Merge<ComponentExplicitOutput<TC<C1>> & ComponentExplicitOutput<TC<C2>>>,
+  Merge<ComponentExplicitOutput<TC<C1>> & ComponentExplicitOutput<TC<C2>>>
+>;
+
+// prettier-ignore
+export type ArrayToComponent<A extends Array<Child>> =
+  A extends [CE] ? TC<A[0]> :
+  A extends [CE, CE] ? MC<A[0], A[1]> :
+  A extends [CE, CE, CE] ? MC<A[0], MC<A[1], A[2]>> :
+  A extends [CE, CE, CE, CE] ? MC<A[0], MC<A[1], MC<A[2], A[3]>>> :
+  A extends [CE, CE, CE, CE, CE] ? MC<A[0], MC<A[1], MC<A[2], MC<A[3], A[4]>>>> :
+  A extends [CE, CE, CE, CE, CE, CE] ? MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], A[5]>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], A[6]>>>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], MC<A[6], A[7]>>>>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], MC<A[6], MC<A[7], A[8]>>>>>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], MC<A[6], MC<A[7], MC<A[8], A[9]>>>>>>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], MC<A[6], MC<A[7], MC<A[8], MC<A[9], A[10]>>>>>>>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], MC<A[6], MC<A[7], MC<A[8], MC<A[9], MC<A[10], A[11]>>>>>>>>>>>> :
+  A extends [CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE, CE] ? TC<MC<A[0], MC<A[1], MC<A[2], MC<A[3], MC<A[4], MC<A[5], MC<A[6], MC<A[7], MC<A[8], MC<A[9], MC<A[10], MC<A[11], A[12]>>>>>>>>>>>>> :
```

Comment:
That's a lot of `>`s in succession :astonished: But, I don't think there is a better way to do it.